### ISSUE TITLE
Android: `deletePhotos` not working with `content://` assets

### ISSUE THREAD

**Issue by mrousavy** [react-native-cameraroll/react-native-cameraroll#639]
On Android, I can't seem to be able to delete photos in the user's gallery.

I have the following Asset/`PhotoIdentifier`:

```json
{
  "group_name": [
    "Pictures"
  ],
  "id": "299",
  "image": {
    "extension": "jpg",
    "fileSize": 1065485,
    "filename": "mrousavy-3219287467027880592.jpg",
    "height": 1920,
    "orientation": 0,
    "playableDuration": null,
    "uri": "content://media/external/file/299",
    "width": 1080
  },
  "location": null,
  "modificationTimestamp": 1718290576,
  "sourceType": "UserLibrary",
  "subTypes": [],
  "timestamp": 1718290575.821,
  "type": "image/jpeg"
}
```

So it's `image.uri` is `content://media/external/file/299` - I use this code to delete it:

```ts
await CameraRoll.deletePhotos([asset.node.image.uri])
```

When I try to delete that, it fails with this error:

```diff
- Could not delete all media, only deleted 0 photos.
```

Any idea why?

**Comment by mrousavy** [comment]
cc @Arjan-Zuidema @bartolkaruza have you seen this before? sorry for pinging, just trying to figure out if this is my fault or if it's a bug I can fix in the library :)

**Comment by Arjan-Zuidema** [comment]
Not seen that before, but does not mean it's not a bug. The file exists but you cannot delete? Could it be a permission issue? They changed permissions on Android a lot lately :')

**Comment by mrousavy** [comment]
Oh yeah true I'll double check if you need special permission to do that.

If yes, I can add some code that checks for that and gives a better error message! :)

**Comment by kaiowz** [comment]
I was having the same problem. I followed the pattern that the documentation states and resolved it this way: 
` let uri = `file:///storage/emulated/0/DCIM/inspecao-pro/${photo.filename}`;
CameraRoll.deletePhotos([uri])
        `

**Comment by mrousavy** [comment]
That doesn't really seem safe, the path can be different from the hardcoded prefix.. Maybe we forget to expose the actual native part on the Java side?

**Comment by Romick2005** [comment]
I've got similar issue, but in my case I was trying to upload image, but the path provided on uri does not exist.
Here are my results on Android 14 device:
```
onMediaSelectionDone [
  {
    "type": "image",
    "uri": "content://media/external/file/1000004215",
    "width": 1630,
    "height": 1920,
    "filename": "IMG_6516.JPG",
    "fileSize": 2947839,
    "playableDuration": null
  }
]

RNFS.stat(filePath) filePath = content://media/external/file/1000004215; 
fileStat: {
  "path": "content://media/external/file/1000004215",
  "ctime": "2024-06-10T15:00:44.000Z",
  "mtime": "2024-06-10T15:00:44.000Z",
  "size": 2947839,
  "originalFilepath": "/storage/emulated/0/DCIM/APP_NAME/IMG_6516.JPG"
}

RNFS.exists(filePath) filePath = content://media/external/file/1000004215; exists: false
RNFS.exists(originalFilepath) originalFilepath = /storage/emulated/0/DCIM/APP_NAME/IMG_6516.JPG; exists: true
```

**Comment by shahzaib14944** [comment]
@Romick2005 I tried to delete using the orignalPath but it still gave error on android, like com.appName doesn't have permission to delete content://...

**Comment by Romick2005** [comment]
Original path will not contain "content" prefix. You need to get original path from content* path and only then delete original path.

**Comment by shahzaib14944** [comment]
@Romick2005 check this issue https://stackoverflow.com/questions/73145329/cameraroll-deletephotosuri-react-native-community-cameraroll-com-om-has-no
I'm passing the orignalPath as [uri] but it seems to be converting that uri again to content:// I guess. I think something's wrong with this library.

**Comment by Romick2005** [comment]
Can you please provide some example with data? Also what RNFS.stat(contentFilePath) return? Where RNFS is react-native-fs module.

**Comment by shahzaib14944** [comment]
@Romick2005

**Comment by shahzaib14944** [comment]
@Romick2005 

 LOG  content://media/external/file/1000001074 file uri before stat
 LOG  content://media/external/file/1000001075 file uri before stat
 LOG  content://media/external/file/1000001073 file uri before stat
 LOG  content://media/external/file/1000001076 file uri before stat
 LOG  content://media/external/file/1000001077 file uri before stat
 
 LOG  /storage/emulated/0/DCIM/Camera/PXL_20240730_162815120.jpg url after stat
 LOG  /storage/emulated/0/DCIM/Camera/PXL_20240730_162815520.jpg url after stat
 LOG  /storage/emulated/0/DCIM/Camera/PXL_20240730_162814696.jpg url after stat
 LOG  /storage/emulated/0/DCIM/Camera/PXL_20240730_162815908.MP.jpg url after stat
 LOG  /storage/emulated/0/DCIM/Camera/PXL_20240730_162816291.MP.jpg url after stat
 LOG  ["/storage/emulated/0/DCIM/Camera/PXL_20240730_162815120.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162815520.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162814696.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162815908.MP.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162816291.MP.jpg"] array
 
 now passing this array in  CameraRoll.deletePhotos(array).then(() => {
                    toastLong("Files deleted successfully");
                }).catch((e) => {
                    console.log(e, "error");
                    toastLong("Error deleting files");
                });
                gives this error: 
![error](https://github.com/user-attachments/assets/54d0b5cb-eb43-4363-9fb3-a42a2bc96460)

**Comment by shahzaib14944** [comment]
@Romick2005 

LOG content://media/external/file/1000001074 file uri before stat
LOG content://media/external/file/1000001075 file uri before stat
LOG content://media/external/file/1000001073 file uri before stat
LOG content://media/external/file/1000001076 file uri before stat
LOG content://media/external/file/1000001077 file uri before stat

LOG /storage/emulated/0/DCIM/Camera/PXL_20240730_162815120.jpg url after stat
LOG /storage/emulated/0/DCIM/Camera/PXL_20240730_162815520.jpg url after stat
LOG /storage/emulated/0/DCIM/Camera/PXL_20240730_162814696.jpg url after stat
LOG /storage/emulated/0/DCIM/Camera/PXL_20240730_162815908.MP.jpg url after stat
LOG /storage/emulated/0/DCIM/Camera/PXL_20240730_162816291.MP.jpg url after stat
LOG ["/storage/emulated/0/DCIM/Camera/PXL_20240730_162815120.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162815520.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162814696.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162815908.MP.jpg", "/storage/emulated/0/DCIM/Camera/PXL_20240730_162816291.MP.jpg"] array

now passing this array in CameraRoll.deletePhotos(array).then(() => {
toastLong("Files deleted successfully");
}).catch((e) => {
console.log(e, "error");
toastLong("Error deleting files");
});
gives this error:
![error](https://github.com/user-attachments/assets/0564ce87-b71a-4924-a1f2-9410418e4127)

**Comment by Romick2005** [comment]
So when you got original real path you can call RNFS.unlink to remove file. Smth like:
RNFS.unlink("/storage/emulated/0/DCIM/Camera/PXL_20240730_162815120.jpg") in your case, but not the path with "content://"

**Comment by shahzaib14944** [comment]
@Romick2005 I already tried working with RNFS unlink but it didn't work as expected due to android scope storage permission. It shows success in response but fails to delete. Also tried this package https://www.npmjs.com/package/react-native-delete-media but seems like it's not updated to support android 13 and 14 as WRITE_EXTERNAL_STORAGE is no longer required in API lvl >= 33 and library fails due to that.

**Comment by Romick2005** [comment]
Works for me

**Comment by mrousavy** [comment]
Well, I'd like to avoid installing an additional dependency (`react-native-fs`) just to delete a photo. I think this is a bug report, and should be fixed in CameraRoll.

I can take a look at fixing this soon.

**Comment by dws-krisha** [comment]
CameraRoll.deletePhotos([uri]) in iOS working fine but in Android getting error Error deleting photo: [Error: Could not delete all media, only deleted 0 photos.] due to file path issue.
Does this issue fixed?

**Comment by arnabJ** [comment]
> @Romick2005 I tried to delete using the orignalPath but it still gave error on android, like com.appName doesn't have permission to delete content://...

I can confirm the same. On my Android 14 Emulator, after converting the content:// uri to file:// uri, I got this same error.

**Comment by Romick2005** [comment]
Can you please try on real device? Also please add some logs what output did you get for the paths and the error message itself.

**Comment by arnabJ** [comment]
> Can you please try on real device? Also please add some logs what output did you get for the paths and the error message itself.

I will try to do this during the weekend. Stuck with other projects as of now.

**Comment by dws-krisha** [comment]
@Romick2005  below is my function to delete image
  
  const resolveUriToFilePath = async (uri: string): Promise<string> => {
    if (Platform.OS === 'android' && uri.startsWith('content://')) {
      const filePath = await RNFetchBlob.fs.stat(uri);
      console.log('FILE_CONTENT_PATH', filePath);
      return filePath.path; // Return the resolved file path
    }

    // On iOS, or if the URI is already a file path
    else {
      return uri.replace('file://', '');
    }
  };

  const deletePhoto = async (uri: string) => {
    try {
      if (Platform.OS === 'android') {
        const filePath = await resolveUriToFilePath(uri);
        console.log('Resolved File Path:', filePath);
        await RNFS.unlink('file://' + filePath); // Delete the file from Android filesystem
        console.log('DELETED_FILE_PATH', filePath);
      } else if (Platform.OS === 'ios') {
        await CameraRoll.deletePhotos([uri]); // Delete the photo from iOS Photos library
      }

      Alert.alert(
        'Photo deleted successfully',
        'Would you like to refresh the photos list?',
        [
          {
            text: 'Cancel',
            style: 'cancel',
          },
          {
            text: 'OK',
            onPress: () => {
              getTotalImageCount();
              fetchPhotos(); // Call fetchPhotos to reload the photo list
            },
          },
        ],
        {cancelable: false},
      );

      // Update the list after deleting the image
      setSimilarPhotos(prevPhotos =>
        prevPhotos.filter(
          photo => photo.original !== uri && photo.duplicate !== uri,
        ),
      );
    } catch (error) {
      console.error('Error deleting photo:', error);
      Alert.alert('Failed to delete photo');
    }
  };
  
  And here are logs i have added
   FILE_CONTENT_PATH {"filename": "IMG_20230807_172955.jpg", "lastModified": 1691409752000, "path": "/storage/emulated/0/DCIM/IMG_20230807_172955.jpg", "size": 60915, "type": "file"}
 LOG  Resolved File Path: /storage/emulated/0/DCIM/IMG_20230807_172955.jpg
 LOG  DELETED_FILE_PATH /storage/emulated/0/DCIM/IMG_20230807_172955.jpg
 
 I got image deleted successfully but it is not getting deleted from gallery and when i again fetch the images list from gallery i got error  Error converting image to base64 [Error: ENOENT: open failed: ENOENT (No such file or directory), open 'content://media/external/file/7293'] due to file not deleted succesfully.
 
 Any idea how to resolve this?

**Comment by Romick2005** [comment]
As you may noticed your link does not have 'content://' or 'file://' prefix. So if RN cameraroll provide you with a file that have "content://" as a prefix only then you can get real path and remove file. Also it might not work on Android emulator, at least I did't test it on emulator. One small note if you try to remove Android file you probably will need 'file://' prefix in the path.

**Comment by shahzaib14944** [comment]
> > @Romick2005 I tried to delete using the orignalPath but it still gave error on android, like com.appName doesn't have permission to delete content://...
> 
> I can confirm the same. On my Android 14 Emulator, after converting the content:// uri to file:// uri, I got this same error.

I developed a native module (with the help of GPT) that implements the MediaStore.createDeleteRequest, which prompts the user for confirmation, similar to the behavior in iOS. It worked flawlessly. Deleting files using RNFS is not possible, especially on Android 13 and later due to scoped storage restrictions.

One adjustment I made to ensure my native module worked was handling the URI format returned by the CameraRoll library. For example, the library provides URIs in the format content://media/external/file/1000001077, but I had to convert them to content://media/external/images/media/1000001077. I achieved this by replacing the file path like so:

let uri = file?.uri?.replace("file", "images/media");

Perhaps you could implement this URI format conversion directly within the CameraRoll library and see if it enables file deletion. It worked with my native module, though I haven't tested it with this library specifically.

**Comment by arnabJ** [comment]
@shahzaib14944 I also tried using Chat GPT for this purpose, but no matter how many times I tried, it always failed to delete.

Can you please share your Native Module code?

**Comment by mrousavy** [comment]
Is the native module implemented with [Nitro](https://mrousavy.github.io/nitro/)? 😄

**Comment by shahzaib14944** [comment]
Hey Arjan, I'm working on it in the little free time I can find. I'm struggling a bit with the native code but have still made some progress. I'll let you know once it's done—though you'll still need to find time to refactor the code anyway! 😄

**Comment by Zeeshan404** [comment]
- Android 9 and below we need `READ_EXTERNAL_STORAGE` and `WRITE_EXTERNAL_STORAGE` storage permissions.
- Android 10 we can set a flag in manifest to `requestLegacyExternalStorage="true`" to opt out of scope storage.
- Android 11 or higher, now we have no way to delete directly either use `createDeleteRequest` from `MediaStore` as @shahzaib14944 implemented. Not a good as it pops up confirmation every time.
- If you want full access and no confirmation pop ups you have to use `MANAGE_EXTERNAL_STORAGE`(Very difficult to get approval from google) with `MANAGE_MEDIA`.

Moreover, `WRITE_EXTERNAL_STORAGE` is deprecated on Android 10.

These privacy friendly storage changes have made it too hard for developers to catch up.

**Comment by Arjan-Zuidema** [comment]
This should be fixed in version 7.8.4
